### ISSUE TITLE
Segmentation fault (core dumped)  when `lr` is high.

### ISSUE THREAD

**Issue by eromoe** [facebookresearch/fastText#239]
Hi, 

I met this error today, while using `-lr 0.5` :

```
root@a934a7b965b6:/data/fastText# ./fasttext skipgram -input /data/thulac_new.txt -lr 0.5 -epoch 5 -output /data/w2v_model
Read 466M words
Number of words:  463102
Number of labels: 0
Progress: 0.0%  words/sec/thread: 34495  lr: 0.499885  loss: 30.873806  eta: 1h34m 
Segmentation fault (core dumped)
```

But `-lr 0.1` is fine:

```
root@a934a7b965b6:/data/fastText# ./fasttext skipgram -input /data/thulac_new.txt -lr 0.1 -epoch 5 -output /data/w2v_model
Read 466M words
Number of words:  463102
Number of labels: 0
Progress: 0.3%  words/sec/thread: 83157  lr: 0.099717  loss: 1.656397  eta: 0h38m
```
Training data is 2.47GB Chinese text

**Comment by iymitchell** [comment]
I think you are out of memory. I tried a learning rate of 1 on my machine and it worked. 

My machine specs are: 32 Cores 400GB RAM

**Comment by cpuhrsch** [comment]
Hello @eromoe,

I have not been able to reproduce an error like this / all our internal tests run clear. Maybe you can try to reproduce this on one of the open-source dataset that we supply as part of the example scripts. This way we can make sure we're running under the exact same conditions.

Thanks,
Christian

**Comment by EmilStenstrom** [comment]
I get this same error. 

```
$ fasttext supervised -input wikis/wikipedia-classification-short.txt -output wikipedia-classification-short.bin -lr 1.0 -epoch 25 -wordNgrams 2 -bucket 200000 -dim 100 -loss hs
Read 226M words
Number of words:  2445360
Number of labels: 3787938
Progress: 0.8%  words/sec/thread: 257609  lr: 0.992496  loss: 27.614244  eta: 0h30m Segmentation fault: 11
```

While reading the words and labels the memory stays under 1 Gb, but as soon as training starts it spikes above my 16 Gb and crashes.

**Dataset**: The dataset I'm using is wikipedia in swedish, with the title of the page as label, and the first 100 words as text. 

**Computer**: MacBook Pro (Retina, 15 tum, mitten 2014), 2,2 GHz Intel Core i7, 16 GB 1600 MHz DDR3

**Comment by EmilStenstrom** [comment]
Hi, I get this same error. Let me know if you need something else to reproduce.

```
$ fasttext supervised -input wikis/wikipedia-classification-short.txt -output wikipedia-classification-short.bin -lr 1.0 -epoch 25 -wordNgrams 2 -bucket 200000 -dim 100 -loss hs
Read 226M words
Number of words:  2445360
Number of labels: 3787938
Progress: 0.8%  words/sec/thread: 257609  lr: 0.992496  loss: 27.614244  eta: 0h30m Segmentation fault: 11
```

While reading the words and labels the memory stays under 1 Gb, but as soon as training starts ("Progress: 0.1%" shows up) memory spikes to somewhere near 16 Gb and fasttext crashes with Segmentation fault 11.

**Version**: Latest version of fastText from here (git pull) says everything is up to date. Tried make clean and make again just to make sure I really run the latest version. 

**Dataset**: The dataset I'm using is wikipedia in swedish, with the title of the page as label, and the first 100 words as text. So each line looks like this:

```
__label__Barack_Obama Barack Hussein Obama II [bəˈrɑːk hʊˈseɪn obɑːmə] (info), född 4 augusti 1961 i Honolulu, Hawaii, är en amerikansk demokratisk politiker som v...
```

Computer: MacBook Pro (Retina, 15 tum, mitten 2014), 2,2 GHz Intel Core i7, 16 GB 1600 MHz DDR3. I've tried rebooting between runs to make sure it's not something transient.

**Comment by EmilStenstrom** [comment]
Here is a reproducible testcase with your test data:

```bash
$ wget https://s3-us-west-1.amazonaws.com/fasttext-vectors/cooking.stackexchange.tar.gz && tar xvzf cooking.stackexchange.tar.gz
$ for i in {1..1000}; do cat cooking.stackexchange.txt >> "cooking.inflated.txt"; done
$ fasttext supervised -input cooking.inflated.txt -output cooking -lr 1.0 -epoch 25 -wordNgrams 2 -dim 100 -loss hs
Read 184M words
Number of words:  16568
Number of labels: 736
Progress: Segmentation fault: 11: 1336601  lr: 0.987408  loss: 30.378572  eta: 0h4m 
```

It does not crash on the exact same location when run again:

```bash
$ fasttext supervised -input cooking.inflated.txt -output cooking -lr 1.0 -epoch 25 -wordNgrams 2 -dim 100 -loss hs
Read 184M words
Number of words:  16568
Number of labels: 736
Progress: 1.3%  words/sec/thread: 1342591  lr: 0.987393  loss: 30.347952  eta: 0h4m Segmentation fault: 11
```

**Comment by EmilStenstrom** [comment]
It does **not** crash on smaller datasets, like if I only duplicate the data 100 times instead of 1000.

**Comment by cpuhrsch** [comment]
Hello @EmilStenstrom,

Thank you for the detailed description of your problem!  The segfault does indeed appear to be caused by this very high learning rate and I can reproduce it on my side. When setting a more reasonable learning rate (such as 0.01), the training proceeds and finishes successfully. We'll investigate this issue in more detail.

For what it's worth, related to your first issue, most of the memory usage comes from either the input or output parameter matrix. The input matrix for supervised training will be the number of words plus the size of the bucket, which is 2M by default, times the dimensionality (-dim). So, for your first example, you'll need to allocate one matrix (floats) of size (2445360 + 2000000) x 100. The output matrix will need to be able to produce a prediction for each label. So it would be of size 100 x 3787938. These are rather large. To influence the number of words, set a higher number for the '-minCount' option. To change the number of buckets (bins for hashing of non-words) use the -bucket flag. To influence the number of labels use the -minLabelCount option.

Thanks,
Christian

**Comment by EmilStenstrom** [comment]
@cpuhrsch Glad to be able to help! Hope you'll find a fix. I picked the learning rate from the supervised-tutorial, didn't understand that it was as outrageous :)

As for my other issue (I thought they where the same...): I'll play around with the settings, thanks for the pointers on how the get the memory usage down. An idea: If that math is that simple, maybe a (best effort) check would be a good idea? If not just to improve the error message, and avoid github issues like these...

(On a separate note: there are quick a few pull requests in the queue, some of which improve documentation. Maybe worth it to go through?)

**Comment by cpuhrsch** [comment]
Hello @EmilStenstrom,

It's true that picking the learning rate from the tutorial should be reasonable first step. Ideally this simply won't segfault and then it'll be ok anyway haha. I also agree with your second point on a more principled approach for dealing with memory requirements. As the project grows/matures we'll have more time to work on user friendliness / robustness and fix issues like this.

I've been meaning to work through the pull requests next. Thanks for pointing that out.

I'm going to close this issue for now, please feel free to reopen it at any point if it becomes current again.

Thanks,
Christian

**Comment by ChrisYanY** [comment]
@ Christian .  Had the same error this morning.  Going through the chain seems like a memory problem to me.  Maybe I missed something, but could you explain a bit more why learningRate affects the memory?  Seem to me shrinking the size of dim / bucket would be a good direction to try out?  Does decreasing epoch numbers help?   Many thanks in advance.

-yan

**Comment by nmfisher** [comment]
I ran into a similar issue with hierarchical softmax using the default learning rate. On the dataset I was using, if the base learning rate is too high, the gradients/embedding weight updates can explode, meaning embedding matrix elements can overflow the 32 bit float. 

I suspect this is causing the binary logistic score to equal inf (or NaN), so when [x is passed to the sigmoid function](https://github.com/facebookresearch/fastText/blob/e39dfbe598542fbb8a8e25be596ecab3ad25edd9/src/model.cc#L336), the array index is invalid, causing the seg fault.

May be worthwhile checking for inf/NaN and warning the user of an overflow - I think this is the approach taken in the various sklearn models.

**Comment by cpuhrsch** [comment]
Reopening this thread due to continued interest.

**Comment by cpuhrsch** [comment]
Hello @eromoe,

This has been resolved as of 4697db0. I also added this to the FAQ. Please feel free to reopen this at any time, if it's still an issue.

Thanks,
Christian

**Comment by atefeh-f** [comment]
hi,
my system have core i5 and 4 GB RAM. i run a file of lua on torch library. my code about deep learning.
please help me. my error is:
==> processing options  
==> executing all  
data/toy//train.csv  
Gene Count:  10  
Number of entries:  1000  
Number of features:  5  
Read entries!!  
data/toy//valid.csv  
Gene Count:  10  
Number of entries:  1000  
Number of features:  5  
Read entries!!  
==>DeepChrome Classification Model  
==> define parameters  
==> construct model  
==> here is the model:  
nn.Sequential {
  [input -> (1) -> (2) -> (3) -> (4) -> (5) -> (6) -> (7) -> (8) -> (9) -> (10) -> output]
  (1): nn.TemporalConvolution
  (2): nn.ReLU
  (3): nn.TemporalMaxPooling
  (4): nn.View(900)
  (5): nn.Dropout(0.500000)
  (6): nn.Linear(900 -> 625)
  (7): nn.ReLU
  (8): nn.Linear(625 -> 125)
  (9): nn.ReLU
  (10): nn.Linear(125 -> 2)
}
==> define loss  
==> here is the loss function:  
nn.ClassNLLCriterion
==> loading data  
==> defining some tools  
==> configuring optimizer : SGD  
==> defining training procedure  
==> defining test procedure  
==> training!  
==> doing epoch on training data:  
Segmentation fault (core dumped)................] ETA: 0ms | Step: 0ms
